### PR TITLE
Update Readme file with GitOps Flow and Build/Deploy Flow

### PR DESCRIPTION
**Pull request by mundrapiyush36**
### Description

Updated README with additional steps for
- GitOps Flow
- Modify Build/Deploy Tasks in the CI/CD Pipelines
- Added ToC
---

### Testing Done

---

### CODE REVIEW

**Comment by akshaykumar-vijapur** [comment]
@mundrapiyush36  We can do a force push to the branch to avoid blank `.` commit messages.

File: README.md [hunk 1]
```diff
@@ -1,142 +1,263 @@
...
+**Note:**
+
+If you are providing Username and Password, enable **PasswordAuthentication** in the Virtual Server Instance.
+
+  
 
 ### Advanced User Scenarios:
 
-Any change to the source triggers the CI Pipeline. The result of a successful CI Run is a new build/binary artifact which is stored first to an intermediate storage and then deployed to the target Virtual Server Instance. The toolchain provides user with an option to use either IBM Cloud Object Store or Artifactory as an intermediate storage to store the build binaries.
+The Continuous Integration (CI) Pipeline listens to changes in the application source code and triggers a CI Pipeline Run whenever a change is pushed to the application repository. The result of a successful CI Run is a artifact which is pushed to an intermediate storage (IBM Cloud Object Store or Artifactory). The Continuous Deployment (CD) Pipeline copies this artifact to the target Virtual Server Instance and performs application deployment. The toolchain provides user with an option to use either IBM Cloud Object Store (COS) or Artifactory as an intermediate storage to store the build binaries.
```
Comment (akshaykumar-vijapur) [inline README.md]: 1. `this artifact` -> `these artifact`

**Review by akshaykumar-vijapur — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -1,142 +1,263 @@
...
+        - [Modify the Continous Integration Pipeline task](#modify-the-continous-integration-pipeline-task)
+        - [Modify the Continuous Deployment Pipeline task](#modify-the-continous-deployment-pipeline-task)
++ [Learn more](#learn-more)
+
 ### Continuously Deliver a Java Application to the Virtual Server Instance.
 
-By following this tutorial, you can create an open toolchain, and then use the toolchain and DevOps practices to develop a simple "Hello World" Java web application that you deploy to the IBM Cloud® Virtual Server Instance.
+  
+
+By following this tutorial, you can create an open toolchain, and then use the toolchain and DevOps practices to develop a simple Java web application that you deploy to the IBM Cloud® Virtual Server Instance.
```
Comment (padraic-edwards) [inline README.md]: By following this tutorial, you can create an IBM Cloud Continuous Delivery Toolchain and then use the toolchain....

File: README.md (hunk 1 above)
Comment (padraic-edwards) [reply README.md]: The Continuous Integration (CI) Pipeline listens to changes in the application source code and triggers a CI Pipeline Run whenever a change is pushed to the application repository. The result of a successful CI Run is an artifact which is pushed to an intermediate storage). The Continuous Deployment (CD) Pipeline copies this artifact to the target Virtual Server Instance and performs the deployment of the application. The toolchain provides a user with options to use either IBM Cloud Object Store (COS) or Artifactory as an intermediate storage location for binary builds.

File: README.md [hunk 3]
```diff
@@ -1,142 +1,263 @@
...
+The Continuous Integration (CI) Pipeline listens to changes in the application source code and triggers a CI Pipeline Run whenever a change is pushed to the application repository. The result of a successful CI Run is a artifact which is pushed to an intermediate storage (IBM Cloud Object Store or Artifactory). The Continuous Deployment (CD) Pipeline copies this artifact to the target Virtual Server Instance and performs application deployment. The toolchain provides user with an option to use either IBM Cloud Object Store (COS) or Artifactory as an intermediate storage to store the build binaries.
+
+-  **Cloud Object Store (Default)** - The toolchain utilizes IBM Cloud Store (COS) to store transient application artifacts. In case user does not have an instance of IBM COS, toolchain creates one on behalf of user. The toolchain uses this newly created instance of IBM COS to create a bucket which will store the application artifacts. Alternately, user can also configure the toolchain to utilise an existing IBM COS Bucket or [Create Bucket](https://cloud.ibm.com/docs/cloud-object-storage/basics?topic=cloud-object-storage-provision)
```
Comment (padraic-edwards) [inline README.md]: The toolchain utilizes IBM Cloud Object Storage (COS) to store transient application artifacts. If a user does not have an instance of IBM COS, the toolchain will go ahead and create an instance for them. The toolchain uses this newly created instance of IBM COS to create a bucket which will store the application artifacts. Alternatively, a user can also configure the toolchain to utilise an existing IBM COS Bucket or [Create a Bucket](https://cloud.ibm.com/docs/cloud-object-storage/basics?topic=cloud-object-storage-provision)

File: README.md [hunk 4]
```diff
@@ -1,142 +1,263 @@
...
+
+-  **Artifactory (Optional)** - The toolchain also provides the capability to utilize an existing instance of JFrog Artifactory as intermittent storage to store the application artifacts (for ex. jar, exe).
```
Comment (padraic-edwards) [inline README.md]: The toolchain also provides the capability to utilize an existing instance of JFrog Artifactory as intermittent storage location for application artifacts (e.g. jar, exe).

File: README.md [hunk 5]
```diff
@@ -1,142 +1,263 @@
...
+
+Please note:
 
- - **Cloud Object Store (Default)** -  The toolchain leverages the IBM Cloud Store to archive intermediate application build binaries. In case user does not have an instance of IBM Cloud Object Store, the toolchain creates one on behalf of user. The toolchain then creates a bucket to store all the built application binaries. In cases when user intends to use an existing instance of IBM Cloud Store and Bucket, user can configure the toolchain to utilize the same. More instructions on creating an instance of Cloud Object Store can be found [here](https://cloud.ibm.com/docs/cloud-object-storage/basics?topic=cloud-object-storage-provision)
+- If artifactory is configured and integrated then artifactory is used.
```
Comment (padraic-edwards) [inline README.md]: If the Artifactory integration is configured correctly with the required fields filled in then that integration is used when determining what store to use for built artifacts.

File: README.md [hunk 6]
```diff
@@ -1,142 +1,263 @@
...
+
+- If the COS bucket name is configured then corressponding bucket is used.
```
Comment (padraic-edwards) [inline README.md]: If the COS bucket field name is configured then the corresponding bucket is used.

File: README.md [hunk 7]
```diff
@@ -1,142 +1,263 @@
...
+
+- If neither Artifatory nor COS bucket name is configured, then toolchain creates a COS instance in the default resource group. Once a COS Instance is successfully created, toolchain creates a bucket within the newly created COS Instance and use the same for storing artifacts.
```
Comment (padraic-edwards) [inline README.md]: If neither the Artifactory or COS integration are configured, then toolchain creates a COS instance in the user’s default resource group. After a COS instance is successfully created, the pipeline creates a bucket within the newly created COS instance and uses this bucket for storing artifacts.

File: README.md [hunk 8]
```diff
@@ -1,142 +1,263 @@
...
-### Inventory Repository Configurtaion
-1. Inventory repo is used to store the metadata of the builds and artifacts. 
-2. Commit on the Inventory repo triggers the CD pipeline to deploy the artifacts on the VSI and do the acceptance task.
+  
+
+### Inventory Repository Configuration
+
+1. Inventory repo is used to capture the build and artifact metadata.
+
+2. Successful CI Build uploads the artifact to COS/Artifactory and commits the build metadata in JSON Format to Inventory Repository. The CD Pipeline triggers pipeline run to fetch the artifact  from COS/Artifactory and deploys it to the Virtual Server Instance.
```
Comment (padraic-edwards) [inline README.md]: A successful CI build uploads the artifact to COS/Artifactory and commits the build metadata in JSON Format to the Inventory Repository. The CD Pipeline listens for changes in the inventory and triggers a pipeline run to fetch the artifact from COS/Artifactory and deploys that artifact to the Virtual Server Instance.

File: README.md [hunk 9]
```diff
@@ -1,142 +1,263 @@
...
+6.  Push the change to inventory repo. Note: The action triggers the CD Pipeline Run for application deployment.
+```
+git push origin master --force
+```  
+
+#### GitOps Flow  
+
+Simple VSI Toolchain follows the GitOps Workflow model. The model stores build metadata from each successful build in a separate repository (Inventory Repository). The default configuration of Continuous Deployment Pipeline triggers a pipeline run whenever a successful build metadata is pushed to the master branch. 
+
+As best practice, user can create a branch in Inventory Repository for each deployment environment. The latest commit to the branch contains metadata of the artifact version deployed in the corresponding environment. Below steps guides user to create a workflow
```
Comment (padraic-edwards) [inline README.md]: As a best practice, user's can create a branch in the Inventory repository for each deployment environment.

Comment (padraic-edwards) [reply README.md]: The steps below guide a user on how to follow this workflow

File: README.md [hunk 10]
```diff
@@ -1,142 +1,263 @@
...
+
+8. Click on the Merge Icon. This action will trigger the CD pipeline run which is listening for changes on prod branch of the Inventory Repository.
+![Submit Merge Request](https://github.com/open-toolchain/simple-vsi-toolchain/blob/master/.bluemix/docs-images/GitOps-7.png) 
+
+9. Verify the CD Pipeline Run. 
+![Submit Merge Request](https://github.com/open-toolchain/simple-vsi-toolchain/blob/master/.bluemix/docs-images/GitOps-8.png) 
+
+#### Customise the build and deploy script
+
+Simple VSI Toolchain leverages tekton perform the pipeline tasks for the build and deploy pipelines. Users can configure the build and deploy steps by making changes to the Pipeline Code which is maintained in the Pipeline Repository.
```
Comment (padraic-edwards) [inline README.md]: The simple VSI toolchain leverages `Tekton` to perform the pipeline tasks for the build and deploy pipelines. Users can configure the build and deploy steps by making changes to the pipeline code which is maintained in the pipeline repository.

File: README.md [hunk 11]
```diff
@@ -1,142 +1,263 @@
...
+9. Verify the CD Pipeline Run. 
+![Submit Merge Request](https://github.com/open-toolchain/simple-vsi-toolchain/blob/master/.bluemix/docs-images/GitOps-8.png) 
+
+#### Customise the build and deploy script
+
+Simple VSI Toolchain leverages tekton perform the pipeline tasks for the build and deploy pipelines. Users can configure the build and deploy steps by making changes to the Pipeline Code which is maintained in the Pipeline Repository. 
+
+##### Modify the Continous Integration Pipeline task
+
+The CI Pipeline is responsible for compiling the application source code, running unit test and other tools. Users can modify the existing tasks for the changing needs of their application. The source code for CI Pipeline is present in **simple-vsi-app-toolchain-\* > .build_pipeline** folder.
```
Comment (padraic-edwards) [inline README.md:235]: The CI Pipeline is responsible for compiling the application source code, running unit tests and other checks. Users can modify the existing tasks for the changing needs of their application. The source code for CI Pipeline is present in **simple-vsi-app-toolchain-\* > .build_pipeline** folder.

File: README.md (hunk 1 above)
Comment (mundrapiyush36) [reply README.md]: Done.

File: README.md (hunk 2 above)
Comment (mundrapiyush36) [reply README.md]: Done.

File: README.md (hunk 3 above)
Comment (mundrapiyush36) [reply README.md]: Done.

File: README.md (hunk 4 above)
Comment (mundrapiyush36) [reply README.md]: Done.

File: README.md (hunk 5 above)
Comment (mundrapiyush36) [reply README.md]: Done.

File: README.md (hunk 6 above)
Comment (mundrapiyush36) [reply README.md]: Done.

File: README.md (hunk 7 above)
Comment (mundrapiyush36) [reply README.md]: Done.

File: README.md (hunk 8 above)
Comment (mundrapiyush36) [reply README.md]: Done.

File: README.md (hunk 9 above)
Comment (mundrapiyush36) [reply README.md]: Done.

File: README.md (hunk 10 above)
Comment (mundrapiyush36) [reply README.md]: Done.

File: README.md (hunk 11 above)
Comment (mundrapiyush36) [reply README.md:235]: Done.

**Pull request merged**